File: java/src/main/java/com/ngt/window/CountWindowAllDemo.java
package com.ngt.window;

import org.apache.flink.configuration.Configuration;
import org.apache.flink.streaming.api.datastream.AllWindowedStream;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.windowing.windows.GlobalWindow;

/**
 * @author ngt
 * @create 2021-01-28 17:16
 * 不分组划分窗口，按照数据的条数划分窗口，属于GlobalWindow
 */
public class CountWindowAllDemo {
    public static void main(String[] args) throws Exception {
        Configuration conf = new Configuration();
        conf.setInteger("rest.port", 8181);
        StreamExecutionEnvironment env = StreamExecutionEnvironment.createLocalEnvironmentWithWebUI(conf);

        DataStreamSource<String> lines = env.socketTextStream("192.168.31.8", 8888);
        SingleOutputStreamOperator<Integer> nums = lines.map(Integer::parseInt);

        // 不分组，将整体当成一个组。并行度为1
        AllWindowedStream<Integer, GlobalWindow> window = nums.countWindowAll(4, 2);

        // 增量聚合：累加是来一条数据加一条，因为只有一个分区能做到增量聚合，同时也避免大量保存数据的空间浪费
        window.sum(0).print();
        env.execute();
    }
}

File: scala/src/main/scala/com/ngt/window/CountWindowAllDemo.scala
package com.ngt.window

import org.apache.flink.streaming.api.scala._
import org.apache.flink.streaming.api.windowing.windows.GlobalWindow

/**
 * @author ngt
 * @create 2021-02-03 20:15
 */
object CountWindowAllDemo {
  def main(args: Array[String]): Unit = {
    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
    val lines: DataStream[String] = env.socketTextStream("192.168.31.8", 8888)

    val nums: DataStream[Int] = lines.map(_.toInt)
    // 将输入的数据每 4 个一组，然后2个滚动一次求和
    val window: AllWindowedStream[Int, GlobalWindow] = nums.countWindowAll(4, 2)
    window.sum(0).print()
    env.execute()
  }
}

/*
3
4  →  7
5
2  →  14
8
9  →  24
注意当输入两个数据的时候就会输出结果
 */